src/kinetics/Reaction.cpp

```diff
@@ -891,6 +825,329 @@ void BlowersMaselInterfaceReaction::getParameters(AnyMap& reactionNode) const
     }
 }
 
+Reaction3::Reaction3(const Composition& reactants, const Composition& products)
+    : Reaction(reactants, products)
+{
+}
+
+void Reaction3::setParameters(const AnyMap& node, const Kinetics& kin)
+{
+    if (!node.hasKey("equation")) {
+        // empty node: used by newReaction() factory loader
+        return;
+    }
+
+    parseReactionEquation(*this, node["equation"], kin);
+    // Non-stoichiometric reaction orders
+    if (node.hasKey("orders")) {
+        for (const auto& order : node["orders"].asMap<double>()) {
+            orders[order.first] = order.second;
+            if (kin.kineticsSpeciesIndex(order.first) == npos) {
+                setValid(false);
+            }
+        }
+    }
+
+    // Flags
+    id = node.getString("id", "");
+    duplicate = node.getBool("duplicate", false);
+    allow_negative_orders = node.getBool("negative-orders", false);
+    allow_nonreactant_orders = node.getBool("nonreactant-orders", false);
+
+    calculateRateCoeffUnits(kin);
+    input = node;
+}
+
+std::pair<std::vector<std::string>, bool> Reaction3::undeclaredThirdBodies(
+        const Kinetics& kin) const
+{
+    std::vector<std::string> undeclared;
+    if (m_third_body) {
+        updateUndeclared(undeclared, m_third_body->efficiencies, kin);
+        bool specified_collision_partner = dynamic_cast<const ThreeBodyReaction3*>(
+            this)->specified_collision_partner;
+        return std::make_pair(undeclared, specified_collision_partner);
+    }
+    return std::make_pair(undeclared, false);
+}
+
+void Reaction3::validate()
+{
+    Reaction::validate();
+    m_rate->validate(equation());
+}
+
+ElementaryReaction3::ElementaryReaction3()
+    : Reaction3()
+{
+    m_rate = std::shared_ptr<ArrheniusRate>(new ArrheniusRate);
+}
+
+ElementaryReaction3::ElementaryReaction3(const Composition& reactants,
+                                         const Composition& products,
+                                         const ArrheniusRate& rate)
+    : Reaction3(reactants, products)
+{
+    m_rate = std::make_shared<ArrheniusRate>(rate);
```

Comment:
The following would make it more clear that this is invoking the copy constructor.
```suggestion
    m_rate.reset(new ArrheniusRate(rate));
```